R/data.table.R

```diff
@@ -1885,36 +1885,37 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
            " but should be nrow(x)==", nrow(x))
   }
   if (!is.null(rownames)) {
-    # extract that column and drop it.
+    # extract that column.
     rownames.value = x[[rownames]]
-    dm = dim(x) - 0:1
+    dm = dim(x) - 0L:1L
     cn = names(x)[-rownames]
-    X = x[, .SD, .SDcols = cn]
   } else {
     dm = dim(x)
     cn = names(x)
-    X = x
+    rownames = ncol(x) + 1L # if no rownames to drop, set index to > ncol so that we can do [-rownames] rather than if (is.null(rownames)) {} else {} to drop the column in the rest of the code
   }
+  X = x
   if (any(dm == 0L))
     return(array(NA, dim = dm, dimnames = list(rownames.value, cn)))
-  p = dm[2L]
+  p = length(cn)
   n = dm[1L]
   collabs = as.list(cn)
   class(X) = NULL
   non.numeric = non.atomic = FALSE
   all.logical = TRUE
-  for (j in seq_len(p)) {
+  # use setDT to check whether any columns are multi-column (e.g. matrices), then unpack using as.data.table
+  multi.col <- tryCatch({ setDT(x); FALSE }, warning=function(w) { return(TRUE) })
```

Comment:
I would either just use the code from `setDT`:

```
if (length(x) > 1L && any(vapply_1i(x, function(xi) length(dim(xi))) > 1L)) warning(...)
```

Or write a helper to be used in both `setDT` and here, like `has_wide_columns(x)`.

Reply:
Resolved as suggested (with minor modification) in latest commit